File: common/python/LeptonCondition.py
##----------------------------------------------------------------------------##
## PoTATo's a Top Analysis Tool
## Project: common
## File:    LeptonCondition.py
## Author:  Joscha Knolle
## Date:    2018-04-30
##----------------------------------------------------------------------------##

import ROOT

from Condition import Condition

class LeptonCondition(Condition):

    def __init__(self, **options):
        Condition.__init__(self, **options)
        self.setOption(options, 'muons', None)
        self.setOption(options, 'electrons', None)
        self.setMinMax(options, 'Pt', -1.0)
        self.setMinMax(options, 'Eta', -1.0)
        self.setOption(options, 'requireLooseId', False)
        self.setOption(options, 'invertLooseId', False)
        self.setOption(options, 'requireMediumId', False)
        self.setOption(options, 'invertMediumId', False)
        self.setOption(options, 'requireTightId', False)
        self.setOption(options, 'invertTightId', False)
        self.setMinMax(options, 'PfIso', -1.0)
        self.setMinMax(options, 'MiniIso', -1.0)

    def getAction(self):
        if self.muons and not self.electrons:
            cutelec, cutmuon = False, True
        elif self.electrons and not self.muons:
            cutelec, cutmuon = True, False
        else:
            cutelec, cutmuon = True, True
        self.action = self.createCondition(
            ROOT.LeptonCondition,
            self.minPt, self.maxPt, self.minEta, self.maxEta,
            cutelec, cutmuon, self.requireLooseId, self.invertLooseId,
            self.requireMediumId, self.invertMediumId, self.requireTightId, self.invertTightId, 
            self.minPfIso, self.maxPfIso,
            self.minMiniIso, self.maxMiniIso
        )
        return Condition.getAction(self)

File: common/src/LeptonCondition.cpp
//----------------------------------------------------------------------------//
// PoTATo's a Top Analysis Tool
// Project: common
// File:    LeptonCondition.cpp
// Author:  Joscha Knolle
// Date:    2018-04-22
//----------------------------------------------------------------------------//

#include <TMath.h>

#include "LeptonCondition.h"

LeptonCondition::LeptonCondition(
    float minpt,
    float maxpt,
    float mineta,
    float maxeta,
    bool cutelec,
    bool cutmuon,
    bool requireloose,
    bool invertloose,
    bool requiremedium,
    bool invertmedium,
    bool requiretight,
    bool inverttight,
    float miniso,
    float maxiso,
    float minmini,
    float maxmini
) :
    Condition<Lepton>(minpt, maxpt, mineta, maxeta),
    _cutElectrons(cutelec),
    _cutMuons(cutmuon),
    _requireLooseId(requireloose),
    _invertLooseId(invertloose),
    _requireMediumId(requiremedium),
    _invertMediumId(invertmedium),
    _requireTightId(requiretight),
    _invertTightId(inverttight),
    _minIso(miniso),
    _maxIso(maxiso),
    _minMiniIso(minmini),
    _maxMiniIso(maxmini)
{}

std::string LeptonCondition::getName() {
    if(_cutElectrons && !_cutMuons) return "ElectronCondition";
    else if(_cutMuons && !_cutElectrons) return "MuonCondition";
    else return "LeptonCondition";
}

bool LeptonCondition::check(Lepton* lep) {
    // Check lepton flavour
    if(!_cutElectrons && lep->isElectron()) return true;
    if(!_cutMuons && lep->isMuon()) return true;

    // Apply cuts on kinematics
    if(!Condition<Lepton>::check(lep)) return false;

    // Apply cuts
    if(_requireLooseId && _invertLooseId==lep->hasLooseId()) return false;
    if(_requireMediumId && _invertMediumId==lep->hasMediumId()) return false;
    if(_requireTightId && _invertTightId==lep->hasTightId()) return false;
    if((_minIso>0.0 && lep->relIso()<_minIso) || (_maxIso>0.0 && lep->relIso()>_maxIso)) return false;
    if((_minMiniIso>0.0 && lep->miniIso()<_minMiniIso) || (_maxMiniIso>0.0 && lep->miniIso()>_maxMiniIso)) return false;

    return true;
}
